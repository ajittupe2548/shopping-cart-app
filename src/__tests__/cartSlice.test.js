import { configureStore } from "@reduxjs/toolkit";
import cartReducer, {
  addToCart,
  removeFromCart,
  increaseQuantity,
  decreaseQuantity,
  clearCart,
  toggleCart,
} from "../store/cartSlice";

describe("cartSlice", () => {
  let store;

  beforeEach(() => {
    store = configureStore({
      reducer: {
        cart: cartReducer,
      },
    });
  });

  const mockProduct = {
    id: 1,
    title: "Test Product",
    price: 29.99,
    image: "test-image.jpg",
  };

  test("should handle initial state", () => {
    const state = store.getState().cart;
    expect(state.items).toEqual([]);
    expect(state.totalQuantity).toBe(0);
    expect(state.totalAmount).toBe(0);
    expect(state.isOpen).toBe(false);
  });

  test("should handle addToCart", () => {
    store.dispatch(addToCart(mockProduct));
    const state = store.getState().cart;

    expect(state.items).toHaveLength(1);
    expect(state.items[0]).toEqual({
      ...mockProduct,
      quantity: 1,
    });
    expect(state.totalQuantity).toBe(1);
    expect(state.totalAmount).toBe(29.99);
  });

  test("should increase quantity when adding existing product", () => {
    store.dispatch(addToCart(mockProduct));
    store.dispatch(addToCart(mockProduct));
    const state = store.getState().cart;

    expect(state.items).toHaveLength(1);
    expect(state.items[0].quantity).toBe(2);
    expect(state.totalQuantity).toBe(2);
    expect(state.totalAmount).toBe(59.98);
  });

  test("should handle removeFromCart", () => {
    store.dispatch(addToCart(mockProduct));
    store.dispatch(removeFromCart(1));
    const state = store.getState().cart;

    expect(state.items).toHaveLength(0);
    expect(state.totalQuantity).toBe(0);
    expect(state.totalAmount).toBe(0);
  });

  test("should handle increaseQuantity", () => {
    store.dispatch(addToCart(mockProduct));
    store.dispatch(increaseQuantity(1));
    const state = store.getState().cart;

    expect(state.items[0].quantity).toBe(2);
    expect(state.totalQuantity).toBe(2);
    expect(state.totalAmount).toBe(59.98);
  });

  test("should handle decreaseQuantity", () => {
    store.dispatch(addToCart(mockProduct));
    store.dispatch(addToCart(mockProduct)); // quantity = 2
    store.dispatch(decreaseQuantity(1));
    const state = store.getState().cart;

    expect(state.items[0].quantity).toBe(1);
    expect(state.totalQuantity).toBe(1);
    expect(state.totalAmount).toBe(29.99);
  });

  test("should not decrease quantity below 1", () => {
    store.dispatch(addToCart(mockProduct));
    store.dispatch(decreaseQuantity(1));
    const state = store.getState().cart;

    expect(state.items[0].quantity).toBe(1);
    expect(state.totalQuantity).toBe(1);
  });

  test("should handle clearCart", () => {
    store.dispatch(addToCart(mockProduct));
    store.dispatch(clearCart());
    const state = store.getState().cart;

    expect(state.items).toHaveLength(0);
    expect(state.totalQuantity).toBe(0);
    expect(state.totalAmount).toBe(0);
  });

  test("should handle toggleCart", () => {
    store.dispatch(toggleCart());
    let state = store.getState().cart;
    expect(state.isOpen).toBe(true);

    store.dispatch(toggleCart());
    state = store.getState().cart;
    expect(state.isOpen).toBe(false);
  });

  test("should handle removeFromCart for non-existing item", () => {
    const initialState = store.getState().cart;
    store.dispatch(removeFromCart(999)); // Non-existing item
    const state = store.getState().cart;

    expect(state).toEqual(initialState);
  });

  test("should handle increaseQuantity for non-existing item", () => {
    const initialState = store.getState().cart;
    store.dispatch(increaseQuantity(999)); // Non-existing item
    const state = store.getState().cart;

    expect(state).toEqual(initialState);
  });

  test("should handle decreaseQuantity for non-existing item", () => {
    const initialState = store.getState().cart;
    store.dispatch(decreaseQuantity(999)); // Non-existing item
    const state = store.getState().cart;

    expect(state).toEqual(initialState);
  });

  test("matches snapshot with empty cart", () => {
    const state = store.getState().cart;
    expect(state).toMatchSnapshot();
  });

  test("matches snapshot with items in cart", () => {
    store.dispatch(addToCart(mockProduct));
    store.dispatch(addToCart({ ...mockProduct, id: 2, title: "Product 2" }));
    const state = store.getState().cart;
    expect(state).toMatchSnapshot();
  });
});
