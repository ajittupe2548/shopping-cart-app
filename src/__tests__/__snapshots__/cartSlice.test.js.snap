// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`cartSlice matches snapshot with empty cart 1`] = `
{
  "isOpen": false,
  "items": [],
  "totalAmount": 0,
  "totalQuantity": 0,
}
`;

exports[`cartSlice matches snapshot with items in cart 1`] = `
{
  "isOpen": false,
  "items": [
    {
      "id": 1,
      "image": "test-image.jpg",
      "price": 29.99,
      "quantity": 1,
      "title": "Test Product",
    },
    {
      "id": 2,
      "image": "test-image.jpg",
      "price": 29.99,
      "quantity": 1,
      "title": "Product 2",
    },
  ],
  "totalAmount": 59.98,
  "totalQuantity": 2,
}
`;
